design/upload-progress.md

```diff
@@ -0,0 +1,318 @@
+Completed, Failed or PartialFailure
+## Reconciliation of InProgress backups
+
+InProgress backups will not have Backup resource present in the object store.  During reconciliation, backups which
+do not have a Backup resource in the object store will be ignored.
+
+### Removal of InProgress backups from object store
+
+## Plug-in API changes
+
+### UploadProgress struct
+
+    type UploadProgress struct {
+        completed bool                          // True when the operation has completed, either successfully or with a failure
+        err error                               // Set when the operation has failed
+        itemsCompleted, itemsToComplete int64   // The number of items that have been completed and the items to complete
+                                                // For a disk, an item would be a byte and itemsToComplete would be the
+                                                // total size to transfer (may be less than the size of a volume if
+                                                // performing an incremental) and itemsCompleted is the number of bytes
+                                                // transferred.  On successful completion, itemsCompleted and itemsToComplete
+                                                // should be the same
+        started, updated time.Time              // When the upload was started and when the last update was seen.  Not all
+                                                // systems retain when the upload was begun, return Time 0 (time.Unix(0, 0))
+                                                // if unknown.
+    }
+
+### VolumeSnapshotter changes
+
+A new method will be added to the VolumeSnapshotter interface (details depending on plug-in versioning spec)
+
+    UploadProgress(snapshotID string) (UploadProgress, error)
+
+UploadProgress will report the current status of a snapshot upload.  This should be callable at any time after the snapshot
+has been taken.  In the event a plug-in is restarted, if the snapshotID continues to be valid it should be possible to
+retrieve the progress.
+
+`error` is set if there is an issue retrieving progress.  If the snapshot is has encountered an error during the upload,
+the error should be return in UploadProgress and error should be nil.
+
+### BackupItemProgress plug-in
+
+Currently CSI snapshots and the Velero Plug-in for vSphere are implemented as BackupItemAction plugins.  The majority of
+BackupItemAction plugins do not take snapshots or upload data so rather than modify BackupItemAction we introduce two new
+plug-ins, SnapshotItemAction and BackupItemProgress.  SnapshotItemAction will be used in place of BackupItemAction for
+the CSI snapshots and the Velero Plug-in for vSphere and will return a snapshot ID in addition to the item itself.
+BackupItemProgress will work similarly to DeleteItemAction, being called only for item kinds which it
```

Comment:
Is the `BackupItemProgress` tracking the progress of a `BackupItemAction` or `SnapshotItemAction`?
It seems to me that it is tracking the progress of a `SnapshotItemAction`. If that's the case, please consider making this as an API in the `SnapshotItemAction` plugin interface and naming it accordingly. If this assumption is not true, then perhaps we should have a similar API in the `SnapshotItemAction` plugin interface as well. WDYT?

Reply:
If the `BackupItemProgress` is in fact monitoring the progress of a `BackupItemAction`, is it possible to add it as an API in the `BackupItemAction` plugin interface?

Reply:
You're right, it should be SnapshotItemAction, I forgot to change it.
The BackupItemAction/RestoreItemAction/DeleteItemAction split up is the norm right now, so adding another plug-in fits with that model.  Maybe they should both be combined in SnapshotItemAction.

Reply:
OK, I merged them together into SnapshotItemAction